### PR TITLE
Move repeating code from models to utils #270

### LINKED ISSUE #270: Move repeating code from models to utils
Transfer repeating parts of code from models to utils (without changing them)

### PR DESCRIPTION
**Pull request by VargBurz**
closes #270

### CODE REVIEW

File: analytics/analytics/utils/common.py [hunk 1]
```diff
@@ -1,6 +1,10 @@
 import numpy as np
 import pandas as pd
-
+import scipy.signal
```
Comment (jonyrock) [inline analytics/analytics/utils/common.py]: I think we should make split `common.py` to a new module with algorithms. 
Lets just keep it in mind for future.

File: analytics/analytics/utils/common.py [hunk 2]
```diff
@@ -275,4 +279,77 @@ def nan_to_zero(segment, nan_list):
...
+    left_bound = center - window_size
+    right_bound = center + window_size + 1
+    return data[left_bound: right_bound]
+
+def subtract_min_without_nan(segment):
+    if not np.isnan(min(segment)):
+        segment = segment - min(segment)
+    return segment
+
+def get_convolve(segments, av_model, data, window_size):
```
Comment (jonyrock) [inline analytics/analytics/utils/common.py]: I don`t like that you don`t add type declarations for these functions. 
I am interested in types of params and return type.

File: analytics/analytics/utils/common.py [hunk 3]
```diff
@@ -275,4 +279,77 @@ def nan_to_zero(segment, nan_list):
...
+        labeled_segment = utils.get_interval(data, segment, window_size)
+        labeled_segment = utils.subtract_min_without_nan(labeled_segment)
+        auto_convolve = scipy.signal.fftconvolve(labeled_segment, labeled_segment)
+        convolve_segment = scipy.signal.fftconvolve(labeled_segment, av_model)
+        convolve_list.append(max(auto_convolve))
+        convolve_list.append(max(convolve_segment))
+    return convolve_list
+
+
+def find_jump_parameters(segment_data, segment_from_index):
```
Comment (jonyrock) [inline analytics/analytics/utils/common.py]: types...

File: analytics/analytics/utils/common.py [hunk 4]
```diff
@@ -275,4 +279,77 @@ def nan_to_zero(segment, nan_list):
...
+    segment_median, segment_max_line, segment_min_line = utils.get_distribution_density(flat_segment_dropna)
+    jump_height = 0.95 * (segment_max_line - segment_min_line)
+    jump_length = utils.find_jump_length(segment_data, segment_min_line, segment_max_line) # finds all interseprions with median
+    cen_ind = utils.intersection_segment(flat_segment.tolist(), segment_median)
+    jump_center = cen_ind[0]
+    segment_cent_index = jump_center - 5 + segment_from_index
+    return segment_cent_index, jump_height, jump_length
+
+
+def find_drop_parameters(segment_data, segment_from_index):
```
Comment (jonyrock) [inline analytics/analytics/utils/common.py]: types...

File: analytics/analytics/utils/common.py [hunk 5]
```diff
@@ -275,4 +279,77 @@ def nan_to_zero(segment, nan_list):
...
+    segment_median, segment_max_line, segment_min_line = utils.get_distribution_density(flat_segment_dropna)
+    drop_height = 0.95 * (segment_max_line - segment_min_line)
+    drop_length = utils.find_drop_length(segment_data, segment_min_line, segment_max_line)
+    cen_ind = utils.drop_intersection(flat_segment.tolist(), segment_median)
+    drop_center = cen_ind[0]
+    segment_cent_index = drop_center - 5 + segment_from_index
+    return segment_cent_index, drop_height, drop_length
+
+
+def get_distribution_density(segment):
```
Comment (jonyrock) [inline analytics/analytics/utils/common.py]: types...

File: analytics/analytics/utils/common.py [hunk 6]
```diff
@@ -275,4 +279,77 @@ def nan_to_zero(segment, nan_list):
     for val in nan_list:
         segment[val] = 0
     return segment
-    
+
+def find_confidence(segment):
```
Comment (jonyrock) [inline analytics/analytics/utils/common.py]: types...

File: analytics/analytics/utils/common.py [hunk 7]
```diff
@@ -275,4 +279,77 @@ def nan_to_zero(segment, nan_list):
...
+    segment_min = min(segment)
+    segment_max = max(segment)
+    return 0.2 * (segment_max - segment_min)
```
Comment (jonyrock) [inline analytics/analytics/utils/common.py]: lets move this constant to a named constant to top

File: analytics/analytics/utils/common.py [hunk 8]
```diff
@@ -275,4 +279,77 @@ def nan_to_zero(segment, nan_list):
...
         segment[val] = 0
     return segment
-    
+
+def find_confidence(segment):
+    segment_min = min(segment)
+    segment_max = max(segment)
+    return 0.2 * (segment_max - segment_min)    
+
+def get_interval(data, center, window_size):
```
Comment (jonyrock) [inline analytics/analytics/utils/common.py]: types...

File: analytics/analytics/models/drop_model.py [hunk 9]
```diff
@@ -36,78 +36,32 @@ def do_fit(self, dataframe: pd.DataFrame, segments: list) -> None:
...
-                ax_list = list(zip(x, y))
-                ax_list = np.array(ax_list, np.float32)
-                antipeaks_kde = argrelextrema(np.array(ax_list), np.less)[0]
-                segment_median = ax_list[antipeaks_kde[0], 0]
-                cen_ind = utils.intersection_segment(flat_segment.tolist(), segment_median) #finds all interseprions with median
-                drop_center = cen_ind[0] # or -1? test
-                segment_cent_index = drop_center - 5 + segment_from_index
-                deleted_drop = data[segment_cent_index - self.state['WINDOW_SIZE'] : segment_cent_index + self.state['WINDOW_SIZE'] + 1]
-                deleted_drop = deleted_drop - min(labeled_drop)
+                segment_cent_index = utils.find_drop_parameters(segment_data, segment_from_index)[0]
```
Comment (rozetko) [inline analytics/analytics/models/drop_model.py]: What do you get as `[0]` here?

**Pull request merged**